File: basic_tests/topological_sort.py
"""
   Topological Sort
"""

def __getAdjs(node, G_e):
    adjs = []
    for e in G_e:
        if e[0] == node and e[0] not in adjs:
            adjs.append(e[1])
        #elif e[1] == node and e[1] not in adjs:
        #    adjs.append(e[0])

    return adjs

color = None
pi = None
d = None  # time of discovery
f = None  # time of finalization
time = None
found = None

linked_list = None  # to make topological sort

def __start_globals(G_v):
    global color
    global pi
    global d
    global f
    global time
    global found
    
    global linked_list
    
    color = ['WHITE' for v in G_v]
    pi = [None for v in G_v]
    d = [None for v in G_v]
    f = [None for v in G_v]
    time = 0
    found = False
    linked_list = []

def dfs(node2f, G):
    global color
    global pi
    global d
    global f
    global time
    global found
	
    G_v = G[0]
    G_e = G[1]
    
    __start_globals(G_v)
	
    for i in range(0, len(G_v)):
		if color[i] == 'WHITE':
			__dfs_visit(G_v[i], node2f, G_e)
	
    return found

def __dfs_visit(node, node2f, G_e):
	global color
	global pi
	global d
	global f
	global time
	global found
	
	global linked_list
	
	if node == node2f:
		found = True
	
	color[node] = 'GRAY'
	time = time + 1
	
	d[node] = time
	
	print "d= " + str(d)
	
	adj_vs = __getAdjs(node, G_e)
	print "node = %d adjs: " % node,str(adj_vs)
	
	for vl in adj_vs:
		if color[vl] == 'WHITE':
			pi[vl] = node
			__dfs_visit(vl, node2f, G_e)
	
	color[node] = 'BLACK'
	time = time + 1
	f[node] = time
	
	print "f: " + str(f)
	
	linked_list.append(node)


def topological_sort(G):
    for v in G[1]:
		dfs(v, G)
    
    tmp = linked_list[:]
    top_sort = [linked_list.pop() for v in tmp]
    
    print "top_sort: " + str(top_sort)
    
    return top_sort


if __name__ == '__main__':
   #G_e1 = ((8, 9), (3, 7), (1, 4), (7, 8), (0, 5), (5, 2), (3, 8), (2, 9), (0, 6), (4, 9), (2, 6), (6, 4))
   #G_v1 = (0,1,2,3,4,5,6,7,8,9)
   #G1 = (G_v1, G_e1)    
   
   #topological_sort(G1)
   
   print "====================="
   
   G_e2 = ( (0,1), (0,3), (1,2), (3,1), (4,2), (4,5) )
   G_v2 = ( 0, 1, 2, 3, 4, 5 )
   G2 = (G_v2, G_e2)
   
   a2 = topological_sort(G2)
   assert str(a2) == str([4, 5, 0, 3, 1, 2]);
   

File: basic_tests/topological_sort.cpp
#include <cstdio>
#include <cstdlib>
#include <assert.h>
#include <vector>
#include <string>
#include <cstring>
#include <algorithm>

using namespace std;

struct Node {
    int id;
    string color;
    int pi_id;
    int d;  // time of discovery
    int f;  // time of finalization

    vector<int> adjs_nodes_ids;
    
    Node(int identi) {
		this->id = identi;
	}
	
	void set_adjs_nodes(vector<int> adjs) {
		adjs_nodes_ids = adjs;
	}
};

vector<Node> top_sort;
vector<Node> dag;
bool found;
int t;

void __start_globals(vector<Node> G) {
    for(size_t i = 0; i < G.size(); i++) {
		G[i].color = "WHITE";
		G[i].pi_id = -1;
		G[i].d = -1;
		G[i].f = -1;
    }
    
    t = 0;
    found = false;
    dag = G;
}

void __print_dag() {
	printf("--------------------------------\n");
	for(size_t i = 0; i < dag.size(); i++) {
		printf("Node[%d]: ", dag[i].id);
		for(size_t j = 0; j < dag[i].adjs_nodes_ids.size(); j++) {
			printf("%d ", dag[dag[i].adjs_nodes_ids[j]].id);
		}
		printf("\n");
		printf("\td = %d | f = %d | pi_id = %d\n", dag[i].d, dag[i].f, dag[i].pi_id);
	}
}

void __dfs_visit(int node_id, int node2f_id) {
	if (node_id == node2f_id)
		found = true;
	
	dag[node_id].color = "GRAY";
	t++;
	dag[node_id].d = t;
	
	vector<int> adj_nodes_ids = dag[node_id].adjs_nodes_ids;
	
	for (size_t vli = 0; vli < adj_nodes_ids.size(); vli++) {
		if ( dag[adj_nodes_ids[vli]].color.compare("WHITE") == 0 ) {
			dag[adj_nodes_ids[vli]].pi_id = node_id;
			__dfs_visit(adj_nodes_ids[vli], node2f_id);
		}
	}
	
	dag[node_id].color = "BLACK";
	t++;
	dag[node_id].f = t;
	
	__print_dag();

	top_sort.push_back(dag[node_id]);
}

bool dfs(int node2f_id) {
    for(size_t i = 0; i < dag.size(); i++) {
		if((dag[i].color).compare("WHITE") == 0) {
			__dfs_visit(dag[i].id, node2f_id);
		}
	}
	
    return found;
}

void topological_sort(vector<Node> dag) {
    top_sort.clear();
    
    __start_globals(dag);
	__print_dag();
    
    for(size_t i = 0; i < dag.size(); i++) {
		bool b = dfs(dag[i].id);
		printf("dfs[%d]: %s\n", dag[i].id, b? "true" : "false");
	}
	
	printf("top_sort.size: %d\n", (int)top_sort.size());
	
	reverse(top_sort.begin(), top_sort.end());
	
	for(size_t i = 0; i < top_sort.size(); i++) {
		printf("Node: %d\n", top_sort[i].id);
	}
}

int main() {
    vector<Node> dag;
    
    Node n0(0), n1(1), n2(2), n3(3), n4(4), n5(5);
    
    vector<int> adjs0;
    adjs0.push_back(1);
    adjs0.push_back(3);
    n0.set_adjs_nodes(adjs0);
    
    vector<int> adjs1;
    adjs1.push_back(2);
    n1.set_adjs_nodes(adjs1);
    
    vector<int> adjs3;
    adjs3.push_back(1);
    adjs3.push_back(2);
    n3.set_adjs_nodes(adjs3);
    
    vector<int> adjs4;
    adjs4.push_back(2);
    adjs4.push_back(5);
    n4.set_adjs_nodes(adjs4);
    
    dag.push_back(n0);
    dag.push_back(n1);
    dag.push_back(n2);
    dag.push_back(n3);
    dag.push_back(n4);
    dag.push_back(n5);
    
    topological_sort(dag);
    
    assert(top_sort[0].id == 4);
    assert(top_sort[1].id == 5);
    assert(top_sort[2].id == 0);
    assert(top_sort[3].id == 3);
    assert(top_sort[4].id == 1);
    assert(top_sort[5].id == 2);
    
    return 0;
}
